File: mobile/ios/Mesh/Services/MeshWalletCredentials.swift
import Foundation

#if canImport(WalletCore)
import WalletCore

enum MeshWalletCredentials {
    struct Resolved: Equatable {
        let walletID: String
        let address: String
        let importKind: WalletImportKind
        let privateKey: Data
        let mnemonic: [String]?
        let passphrase: String
        let derivationPath: String
    }

    static func resolve(walletID: String? = MeshWalletRegistry.activeWalletID) throws -> Resolved {
        guard let walletID = walletID ?? MeshWalletRegistry.activeWalletID,
              let wallet = MeshWalletRegistry.wallet(id: walletID)
        else {
            throw TronAPIError.broadcastFailed("Wallet is not initialized")
        }

        switch wallet.importKind {
        case .mnemonic:
            guard let words = MeshMnemonicStore.loadWords(walletID: walletID),
                  Valida.allowedMnemonicWordCounts.contains(words.count)
            else {
                throw TronAPIError.broadcastFailed("Wallet is not initialized")
            }
            let passphrase = MeshMnemonicStore.loadPassphrase(walletID: walletID) ?? ""
            let path = TronConfiguration.defaultDerivationPath
            let mnemonic = words.joined(separator: " ")
            let snapshot = try TronWalletService.importWallet(words: words, passphrase: passphrase)
            guard snapshot.address == wallet.address else {
                throw TronAPIError.broadcastFailed("Wallet address mismatch.")
            }
            let privateKey = try TronWalletService.privateKeyData(
                mnemonic: mnemonic,
                passphrase: passphrase,
                derivationPath: path
            )
            return Resolved(
                walletID: walletID,
                address: wallet.address,
                importKind: .mnemonic,
                privateKey: privateKey,
                mnemonic: words,
                passphrase: passphrase,
                derivationPath: path
            )

        case .privateKey:
            guard let hex = MeshPrivateKeyStore.loadHex(walletID: walletID) else {
                throw TronAPIError.broadcastFailed("Wallet is not initialized")
            }
            let privateKey = try TronWalletService.privateKeyData(hex: hex)
            let derived = try TronWalletService.address(fromPrivateKey: privateKey)
            guard derived == wallet.address else {
                throw TronAPIError.broadcastFailed("Wallet address mismatch.")
            }
            return Resolved(
                walletID: walletID,
                address: wallet.address,
                importKind: .privateKey,
                privateKey: privateKey,
                mnemonic: nil,
                passphrase: "",
                derivationPath: ""
            )
        }
    }

    static func supportsHDWalletFeatures(walletID: String? = MeshWalletRegistry.activeWalletID) -> Bool {
        guard let walletID = walletID ?? MeshWalletRegistry.activeWalletID,
              let wallet = MeshWalletRegistry.wallet(id: walletID)
        else { return false }
        return wallet.importKind == .mnemonic
    }

    static func signingKey(
        walletID: String? = MeshWalletRegistry.activeWalletID,
        derivationPath: String? = nil
    ) throws -> Data {
        let resolved = try resolve(walletID: walletID)
        guard let path = derivationPath, !path.isEmpty, path != resolved.derivationPath else {
            return resolved.privateKey
        }
        guard let words = resolved.mnemonic else {
            throw TronAPIError.broadcastFailed("This wallet cannot sign from a derived address.")
        }
        return try TronWalletService.privateKeyData(
            mnemonic: words.joined(separator: " "),
            passphrase: resolved.passphrase,
            derivationPath: path
        )
    }
}
#endif

File: mobile/android/app/src/main/java/com/mesh/wallet/data/MeshWalletCredentials.kt
package com.mesh.wallet.data

import com.mesh.wallet.data.secure.SecureStorage
import com.mesh.wallet.data.secure.WalletRegistry
import com.mesh.wallet.data.tron.TronApiException
import com.mesh.wallet.data.tron.TronWalletService
import com.mesh.wallet.domain.model.WalletImportKind
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet

class MeshWalletCredentials(
    private val registry: WalletRegistry,
    private val secureStorage: SecureStorage
) {
    data class Resolved(
        val walletId: String,
        val address: String,
        val importKind: WalletImportKind,
        val privateKey: ByteArray,
        val mnemonic: List<String>?,
        val derivationPath: String
    )

    fun resolve(walletId: String? = registry.activeWalletId): Resolved {
        val id = walletId ?: registry.activeWalletId
            ?: throw TronApiException.BroadcastFailed("Wallet is not initialized")
        val wallet = registry.wallet(id)
            ?: throw TronApiException.BroadcastFailed("Wallet is not initialized")

        return when (wallet.importKind) {
            WalletImportKind.MNEMONIC -> {
                val words = secureStorage.loadMnemonic(id)
                    ?: throw TronApiException.BroadcastFailed("Wallet is not initialized")
                val snapshot = TronWalletService.importWallet(words)
                require(snapshot.address == wallet.address) {
                    throw TronApiException.BroadcastFailed("Wallet address mismatch.")
                }
                val path = com.mesh.wallet.core.config.TronConfiguration.DEFAULT_DERIVATION_PATH
                val key = signingKeyFromMnemonic(words, path)
                Resolved(id, wallet.address, WalletImportKind.MNEMONIC, key, words, path)
            }
            WalletImportKind.PRIVATE_KEY -> {
                val hex = secureStorage.loadPrivateKey(id)
                    ?: throw TronApiException.BroadcastFailed("Wallet is not initialized")
                val key = TronWalletService.normalizePrivateKeyHex(hex)
                val derived = TronWalletService.importPrivateKey(hex).address
                require(derived == wallet.address) {
                    throw TronApiException.BroadcastFailed("Wallet address mismatch.")
                }
                Resolved(id, wallet.address, WalletImportKind.PRIVATE_KEY, key, null, "")
            }
        }
    }

    fun supportsHdWalletFeatures(walletId: String? = registry.activeWalletId): Boolean {
        val id = walletId ?: return false
        return registry.wallet(id)?.importKind == WalletImportKind.MNEMONIC
    }

    fun signingKey(walletId: String? = null, derivationPath: String? = null): ByteArray {
        val resolved = resolve(walletId)
        if (derivationPath.isNullOrBlank() || derivationPath == resolved.derivationPath) {
            return resolved.privateKey
        }
        val words = resolved.mnemonic
            ?: throw TronApiException.BroadcastFailed("This wallet cannot sign from a derived address.")
        return signingKeyFromMnemonic(words, derivationPath)
    }

    private fun signingKeyFromMnemonic(words: List<String>, path: String): ByteArray {
        val mnemonic = words.joinToString(" ")
        val hdWallet = HDWallet(mnemonic, "")
            ?: throw TronApiException.BroadcastFailed("Invalid mnemonic")
        return hdWallet.getKey(CoinType.TRON, path).data()
    }
}
